### PR TITLE
Add resource `databricks_grant` for managing singular principal

### LINKED ISSUE #2704: [ISSUE] different databricks_grants inside different modules overwrite eachother
Hello,

I have a multi module setup and I'm applying multiple databricks_grants resource blocks (one or more in each module) 

### Configuration

Module A: 

```hcl
resource "databricks_grants" "use_catalog" {
  provider  = databricks.workspace
  catalog   = "strada"
  grant {
    principal  = "Group 1
    privileges = ["USE_CATALOG"]
  }
}
```

Module B:
```hcl
resource "databricks_grants" "use_catalog" {
  provider  = databricks.workspace
  catalog   = "strada"
  grant {
    principal  = "Group 2"
    privileges = ["USE_CATALOG"]
  }
}
```

### Expected Behavior
In this case, databricks_grants.use_catalog in Module A and databricks_grants.use_catalog in Module B have different addresses due to the module prefix. This should ensure that they are treated as separate resources and will not overwrite each other during the apply process.

### Actual Behavior
Every `terraform apply` i run, they keep changing from group 1 to group 2.

### PR DESCRIPTION
**Pull request by martin-walsh**
## Changes
Add resource "databricks_grant" for managing grants for a single principal on a single securable. This allows securing of resources in distributed terraform repos e.g. granting an api user SELECT on a table without wiping out centrally managed privileges such as human user access.

Opening up PR to discuss whether the approach is acceptable to you and will add documentation, and clean up if it is.

Fixes #2704

## Tests

- [x] `make test` run locally
- [x] relevant change in `docs/` folder
- [x] covered with integration tests in `internal/acceptance`
- [x] relevant acceptance tests are passing
- [x] using Go SDK

### CODE REVIEW

**Comment by martin-walsh** [comment]
@Jonathan-Choi FYI

File: catalog/resource_grant.go
Comment (nkvuong) [inline catalog/resource_grant.go:124]: we should let server side validation to kick in

File: catalog/permissions/permissions.go
Comment (nkvuong) [inline catalog/permissions/permissions.go:13]: we should use the go sdk for this, instead of constructing API calls manually

File: catalog/permissions/permissions.go
Comment (nkvuong) [inline catalog/permissions/permissions.go:42]: all these structs are available in go sdk as well

File: catalog/resource_grant.go
Comment (nkvuong) [inline catalog/resource_grant.go:143]: this id won't be unique - we should use the principal in the id

File: catalog/resource_grant.go
Comment (nkvuong) [inline catalog/resource_grant.go:138]: for the new resource, we should use go sdk to avoid tech debt. `databricks_grants` can be refactored later

**Comment by hectorcast-db** [comment]
There is an issue with the current implementation that we are fixing here
https://github.com/databricks/terraform-provider-databricks/pull/3026

This will need to be adjusted here after we merged.

File: catalog/resource_grant.go
Comment (hectorcast-db) [inline catalog/resource_grant.go:87]: We should use `UnityCatalogPrivilegeAssignment` and `common.StructToSchema` , like in the ResourceGrants() resource

File: catalog/resource_grants.go
Comment (hectorcast-db) [inline catalog/resource_grants.go:16]: Nit, move this description

File: catalog/resource_grant.go
Comment (martin-walsh) [reply catalog/resource_grant.go:124]: I matched the impl to that which exists in `resource_grants`. Should I remove it there also for consistency?

File: catalog/resource_grant.go
Comment (martin-walsh) [reply catalog/resource_grant.go:149]: Sure. Matched `resource_grants.go`. Should I adjust there also ?

File: catalog/resource_grant.go
Comment (martin-walsh) [reply catalog/resource_grant.go:87]: Sure, I'll take a look. I have a question though, how would I conditionally mark these fields as `Required` & `ForcesNew` if using the reflective method ?

File: catalog/resource_grant.go
Comment (martin-walsh) [reply catalog/resource_grant.go:149]: Also, having checked the docs, SplitN is at most n strings so it is an extra layer of safety i.e.
```
fmt.Println(strings.SplitN("abc", "/", 2))

Output:
[abc]
```

File: catalog/resource_grants.go
Comment (martin-walsh) [reply catalog/resource_grants.go]: Done

File: catalog/resource_grant.go
Comment (martin-walsh) [reply catalog/resource_grant.go]: Done

File: catalog/resource_grant.go
Comment (martin-walsh) [reply catalog/resource_grant.go]: Done.

File: catalog/resource_grant.go
Comment (martin-walsh) [reply catalog/resource_grant.go:131]: I'm not sure how I can address this in the scope of this PR. The `common.Resource` struct requires `common.DatabricksClient`

**Comment by martin-walsh** [comment]
Added documentation for `databricks_grant` also.

File: catalog/resource_grant.go
Comment (nkvuong) [reply catalog/resource_grant.go:117]: we should leave it as is for `resource_grants` for backwards compatibility, but should avoid it for the new resource

File: catalog/resource_grant.go
Comment (nkvuong) [reply catalog/resource_grant.go:131]: so you can initialise the sdk client using

```go
w, err := c.WorkspaceClient()
if err != nil {
	return err
}

w.Grants.Update(ctx, catalog.UpdatePermissions{
	Changes:       permissionsChange,
	SecurableType: securable,
	FullName:      name,
})
```

Comment (martin-walsh) [reply catalog/resource_grant.go:131]: I made a start (see https://github.com/siteminder-au/terraform-provider-databricks/compare/resource_grant...siteminder-au:terraform-provider-databricks:sdk-update), but I have some concerns that I don't know enough about the backend to make this change. For example, the permissions for shares changes URL & using the new types from the sdk, the StructToSchema generates schema with different Required & Optional settings.

Using the existing `resource_grants.go` as a base, I was comfortable to make the change as it is amounted to little more than verifying that the content of the HTTP requests match what is expected.

I can proceed if you need me to, but I will need some direction

File: docs/resources/grants.md
Comment (nkvuong) [inline docs/resources/grants.md:7]:
```suggestion
Two different resources help you manage your Unity Catalog grants for a securable. Each of these resources serves a different use case:

[databricks_grants](https://registry.terraform.io/providers/databricks/databricks/latest/docs/resources/grants): Authoritative. Sets the grants of a securable and replaces any existing grants defined inside or outside of Terraform.
[databricks_grant](https://registry.terraform.io/providers/databricks/databricks/latest/docs/resources/grant): Authoritative for a given principal. Updates the grants of a securable to a single principal. Other principals within the grants for the securables are preserved.
```

File: catalog/resource_grant.go
Comment (martin-walsh) [reply catalog/resource_grant.go:131]: See the diff above for the `resource_grant_test.go` for the share related tests

File: docs/resources/grant.md
Comment (nkvuong) [inline docs/resources/grant.md:1]: the content of this page would be mostly duplicated from docs/resources/grants.md, so might make sense to just collapse them in a single page, highlighting the differences
wdyt @alexott

File: catalog/resource_grant.go
Comment (martin-walsh) [reply catalog/resource_grant.go:131]: Re the different Optional & Required behaviour, I narrowed it down to the difference of `omitempty` in jsontag
e.g. existing
```go
// PrivilegeAssignment reflects on `grant` block
type PrivilegeAssignment struct {
	Principal  string   `json:"principal"`
	Privileges []string `json:"privileges" tf:"slice_set"`
}
```

sdk
```go
type PrivilegeAssignment struct {
	// The principal (user email address or group name).
	Principal string `json:"principal,omitempty"`
	// The privileges assigned to the principal.
	Privileges []Privilege `json:"privileges,omitempty"`

	ForceSendFields []string `json:"-"`
}
```

Comment (nkvuong) [reply catalog/resource_grant.go:131]: sorry I should have been more specific, for `StructToSchema`, we can leave it to use the same custom struct as `resource_grants` - there are some outstanding work on how we can add additional tf annotation to those structs

it should be easier to swap these API calls
```go
unityCatalogPermissionsAPI := permissions.NewUnityCatalogGrantsAPI(ctx, c)
err := replacePermissionsForPrincipal(unityCatalogPermissionsAPI, securable, name, principal, grants)
```
with the appropriate sdk call (need adjustment for type), e.g.
```go
existing, err := w.Grants.GetBySecurableTypeAndFullName(ctx, catalog.SecurableType(securable), name)
if err != nil {
  return err
}

w.Grants.Update(ctx, catalog.UpdatePermissions{
  Changes:       diffPermissionsForPrincipal(principal, list, existing),
  SecurableType: securable,
  FullName:      name,
})
```

the two resources with special handling here are `shares` (different API call `w.Shares.UpdatePermissions(ctx, sharing.UpdateSharePermissions{})`) and `connection` (reserved TF keyword)

File: catalog/resource_grant.go
Comment (martin-walsh) [reply catalog/resource_grant.go]: Done

File: catalog/resource_grant.go
Comment (martin-walsh) [reply catalog/resource_grant.go:122]: Should I revert the changes to `resource_grants` as the code diverges now ?

Comment (martin-walsh) [reply catalog/resource_grant.go:122]: @nkvuong I reverted those `resource_grants` changes & refactored `resource_grant` to use the SDK. I kept the `UnityCatalogPermissionsAPI` to hide the difference between Shares & Grants APIs.

Comment (martin-walsh) [reply catalog/resource_grant.go:122]: I can bring back the refactor easily enough also & update to use sdk, let me know

Comment (martin-walsh) [reply catalog/resource_grant.go:123]: If needed: https://github.com/siteminder-au/terraform-provider-databricks/compare/resource_grant...siteminder-au:terraform-provider-databricks:resource_grants_refactor

Comment (nkvuong) [reply catalog/resource_grant.go:123]: you can raise another PR for the refactoring of `resource_grants` to Go SDK - makes it easier to review but also revert (if needed)

**Review by nkvuong — COMMENTED** [review]
Thank you again for the contribution - the code looks good to me.

We will need to add some integration tests to this. Since there was no integration test for `databricks_grants`, I'll try to add it to a separate PR first and we can then replicate them for `databricks_grant` in this PR

**Comment by martin-walsh** [comment]
@hectorcast-db I ported over the changes from #3026

**Comment by nkvuong** [comment]
I've added some acceptance tests in #3034, if you can port it to this one to validate

**Comment by martin-walsh** [comment]
> @martin-walsh I've added some acceptance tests in #3034, if you can port it to this one to validate

@nkvuong tests ported

File: docs/resources/grant.md
Comment (nkvuong) [inline docs/resources/grant.md:1]: @martin-walsh since this page is mostly duplicated from grants.md, we should collapse the 2 pages into one grants.md, and highlight the different usage of `databricks_grant` and `databricks_grants`

**Review by nkvuong — APPROVED** [review]
Integration test passed - looks good to me

One last comment about the doc

File: docs/resources/grant.md
Comment (alexott) [reply docs/resources/grant.md:1]: Imho, two pages may help users to understand differences. I think that to avoid duplication we can just have links from another page to common things, like, syntax, etc.

Comment (martin-walsh) [reply docs/resources/grant.md:1]: Do you guys need anything more from me here? I'm not really sure how you guys want your documentation structured.

Comment (nkvuong) [reply docs/resources/grant.md:1]: we should put a link on this page for the content that are the same as `databricks_grants.md`, e.g. line 27 onwards. Something like "See [databricks_grant.md] for more usage details

I've highlighted the part that should be called out differently to `databricks_grants`

File: docs/resources/grant.md
Comment (nkvuong) [inline docs/resources/grant.md:23]:
```suggestion
Terraform will handle any configuration drift on every `terraform apply` run for the specified principal. even when grants are changed outside of Terraform state.
```

File: docs/resources/grant.md
Comment (martin-walsh) [reply docs/resources/grant.md:1]: I can't see the highlight. Do you mean remove everything on `databricks_grants.md` from line 27 onwards ? If so, if it were me I would leave it. It is very user friendly to be able to copy valid syntax.

File: docs/resources/grant.md
Comment (nkvuong) [inline docs/resources/grant.md:325]: I was suggesting to leave the HCL examples here, but defer to grants.md on what permissions can be applied to a resource (as future grants will be added)

File: docs/resources/grant.md
Comment (nkvuong) [inline docs/resources/grant.md:25]:
```suggestion
See [databricks_grants](grants.md) for the list of privilege types that apply to each securable object in the catalog.

## Examples
```

File: docs/resources/grant.md
Comment (martin-walsh) [reply docs/resources/grant.md]: Done

**Review by nkvuong — APPROVED** [review]

File: catalog/permissions/permissions.go
Comment (mgyucht) [inline catalog/permissions/permissions.go:106]: Not for this PR, but a general remark: seems like it would be useful for the SDK itself to expose a function that returns all enum values for an enum type.

File: catalog/resource_grant.go
Comment (mgyucht) [inline catalog/resource_grant.go:139]: Suggestion: for readability, could we make helper methods to construct/deconstruct the ID? It's not obvious what each part of the ID ought to correspond to.

File: catalog/permissions/permissions.go
Comment (mgyucht) [inline catalog/permissions/permissions.go:137]: Is this actually needed? I think you can replace all calls to this function with SetToSlice() and just not calling `.List()` on the parameter.

File: catalog/permissions/permissions.go
Comment (mgyucht) [inline catalog/permissions/permissions.go:144]: This is the same as `slices.DeleteFunc(in, func(s string) { return s == without })`. Can we use the standard library function instead?

File: catalog/permissions/permissions.go
Comment (mgyucht) [inline catalog/permissions/permissions.go:122]: These are nice! I'm going to think about a way to generalize these so you don't have to implement them for every type. This is great for now though.

File: catalog/resource_grant.go
Comment (mgyucht) [inline catalog/resource_grant.go:20]: Maybe I'm misreading this, but because this function only ever operates on one principal, this map will only have one key defined? If so, we can just use `*schema.Set` as opposed to a map. This applies to the other data structures throughout this function.

File: catalog/permissions/permissions.go
Comment (mgyucht) [inline catalog/permissions/permissions.go:69]: Can you comment here that this is based off of `catalog.PermissionsList`? As @nkvuong mentioned, in the future we'll refactor these resources to inherit fields from the Go SDK. Knowing what this is derived from will help us track this down in the future.

File: catalog/resource_grant.go
Comment (mgyucht) [inline catalog/resource_grant.go:152]: Note: here we're calling StructToData on a different datatype than the one that was used in StructToSchema. This works because the catalog.PermissionsList struct contains a subset of the fields of the UnityCatalogPrivilegeAssignment struct, but this could be a bit confusing. Could we return UnityCatalogPrivilegeAssignment from filterPermissionsForPrincipal instead?

File: catalog/resource_grant_test.go
Comment (mgyucht) [inline catalog/resource_grant_test.go:15]: This is really nice that this resource works even if there are existing permissions for a user on the securable.

File: catalog/resource_grant_test.go
Comment (mgyucht) [inline catalog/resource_grant_test.go:261]: Nice!

File: catalog/resource_grant_test.go
Comment (mgyucht) [inline catalog/resource_grant_test.go:361]: Oof... We don't have any control over this error message do we?

File: catalog/resource_grant.go
Comment (mgyucht) [inline catalog/resource_grant.go:18]: Could we rename `pl` to `desired`? Helps with understanding the intent.

**Review by mgyucht — COMMENTED** [review]
Overall, this is a class PR and I support this change. Mainly a couple nits and other remarks but I think with one final pass this is good to merge.

File: catalog/resource_grant.go
Comment (martin-walsh) [reply catalog/resource_grant.go:139]: Matched the existing impl for `resource_grants` as closely as I could

File: catalog/resource_grant.go
Comment (martin-walsh) [reply catalog/resource_grant.go]: Done

File: catalog/permissions/permissions.go
Comment (martin-walsh) [reply catalog/permissions/permissions.go:144]: We could but, DeleteFunc modifies the original slice & the tests fail without cloning the slice first due to side effects. I can swap the code to the following if you prefer:
```suggestion
func SliceWithoutString(in []string, without string) (out []string) {
	out = slices.Clone(in)
	out = slices.DeleteFunc(out, func(s string) bool { return s == without })
	return
}

```

File: catalog/permissions/permissions.go
Comment (martin-walsh) [reply catalog/permissions/permissions.go]: Good catch, pushed

File: catalog/resource_grant.go
Comment (martin-walsh) [reply catalog/resource_grant.go]: Done & added extra defensive check.

File: catalog/resource_grant.go
Comment (martin-walsh) [reply catalog/resource_grant.go]: In any case, added the helper methods to encapsulate the string wrangling

File: catalog/resource_grant_test.go
Comment (martin-walsh) [reply catalog/resource_grant_test.go:361]: Unfortunately not. I did the best I could by using `ConflictsWith` to trim the error down for the `TestResourceGrantCreateOneSecurableOnly` case, but can't do anything here.

File: catalog/permissions/permissions.go
Comment (mgyucht) [reply catalog/permissions/permissions.go:137]: Ah, good point. You can leave this as-is then.

File: catalog/resource_grant.go
Comment (mgyucht) [reply catalog/resource_grant.go:21]: Ping here

**Review by mgyucht — COMMENTED** [review]
This seems great, thanks for addressing my points. Just one question about diffPermissionsForPrincipal, which I think was kind of written to support multiple principals but only ever acts on one; can we clean this up?

**Comment by martin-walsh** [comment]
> This seems great, thanks for addressing my points. Just one question about diffPermissionsForPrincipal, which I think was kind of written to support multiple principals but only ever acts on one; can we clean this up?

It returns an array as that's what the `UpdatePermissions` request takes downstream. It is also a near carbon copy of the impl from resource_grants which to me is a positive as it makes addressing potential bugs easier. I would suggest leaving it as is for those reasons. I have a subsequent PR to update that resource to use go sdk and may refactor further there.

**Comment by mgyucht** [comment]
Gotcha, that makes a lot of sense.

**Review by mgyucht — APPROVED** [review]
Got clarification on the main point. Thanks!

**Comment by mgyucht** [comment]
This failed one of our nightly tests. @nkvuong could you take a look at https://github.com/databricks/eng-dev-ecosystem/actions/runs/7399425920/job/20130815184 and see if you can debug this?

**Comment by nkvuong** [comment]
@mgyucht basically `TestUcAccGrant` and `TestUcAccGrants` are overwriting each other permissions at metastore-level
```hcl
resource "databricks_grant" "metastore" {
	metastore = "{env.TEST_METASTORE_ID}"
	principal  = "%s"
	privileges = ["CREATE_STORAGE_CREDENTIAL"]
}
```
```hcl
resource "databricks_grants" "metastore" {
	metastore = "{env.TEST_METASTORE_ID}"
	grant {
		principal  = "%s"
		privileges = ["CREATE_STORAGE_CREDENTIAL"]
	}
}
```

so either we need to make these 2 tests run sequentially, or remove the metastore bit for `databricks_grant`

**Comment by mgyucht** [comment]
Ah, that makes perfect sense. Can we comment out the databricks_grant test for metastore for now to merge? we need to fix this for metastore assignment for workspaces as well, which also races between two test cases. I can fix both of these at the same time.

File: internal/acceptance/grant_test.go
Comment (nkvuong) [inline internal/acceptance/grant_test.go:63]:
```suggestion
```

**Comment by martin-walsh** [comment]
> Ah, that makes perfect sense. Can we comment out the databricks_grant test for metastore for now to merge? we need to fix this for metastore assignment for workspaces as well, which also races between two test cases. I can fix both of these at the same time.

@mgyucht done.

**Comment by nkvuong** [comment]
@mgyucht integration tests now passed (failed tests are not related to this PR)

**Comment by mgyucht** [comment]
All nightlies passed. Merging now!

@martin-walsh thank you so much for your work on this and your patience through this process! We and our other customers really appreciate contributions like these. This is one of the biggest pain points with permission management for the provider, so you've solved a major problem with this.

**Comment by kvedes** [comment]
Using provider version 1.39, and it seems that `databricks_grant` overwrites current permissions:

```
resource "databricks_grant" "unity_catalog" {
  catalog = "dummy"
  principal  = "group_a"
  privileges = ["CREATE_SCHEMA"]
}

resource "databricks_grant" "unity_catalog2" {
  catalog = "dummy"
  principal  = "group_a"
  privileges = ["USE_CATALOG"]
}
```
This leaves me only with the USE_CATALOG permission.

**Pull request merged**
